pkg/utils/predicate.go

```diff
@@ -21,144 +21,305 @@ import (
+		// that we're not managing since cross namespace ownership is disallowed
+		gto, inset := gtoSet[ref.UID]
+		if !inset {
+			// we have an owner reference to a gittrackobject that isn't in our namespace
+			// be cautious and don't handle this object
+			return false
+		}
+		for _, gtoOwner := range gto.GetOwnerReferences() {
+			if gtoOwner.APIVersion == farosGroupVersion && gtoOwner.Kind == "GitTrack" {
+				return true
 			}
 		}
 	}
 	return false
 }
 
-// NewOwnersOwnerInNamespacePredicate constructs a new OwnersOwnerInNamespacePredicate
-func NewOwnersOwnerInNamespacePredicate(client client.Client) OwnersOwnerInNamespacePredicate {
-	return OwnersOwnerInNamespacePredicate{
-		client:                    client,
-		ownerInNamespacePredicate: NewOwnerInNamespacePredicate(client),
+// OwnersOwnerIsClusterGitTrackPredicate filters events to check the owner's owner of the event
+// object is a ClusterGitTrack
+type OwnersOwnerIsClusterGitTrackPredicate struct {
+	client            client.Client
+	includeNamespaced bool
+}
+
+// NewOwnersOwnerIsClusterGitTrackPredicate constructs a new OwnerIsClusterGitTrackPredicate
+func NewOwnersOwnerIsClusterGitTrackPredicate(client client.Client, includeNamespaced bool) OwnersOwnerIsClusterGitTrackPredicate {
+	return OwnersOwnerIsClusterGitTrackPredicate{
+		client:            client,
+		includeNamespaced: includeNamespaced,
+	}
+}
+
+// Create returns true if the event object owner's owner is a ClusterGitTrack
+func (p OwnersOwnerIsClusterGitTrackPredicate) Create(e event.CreateEvent) bool {
+	return p.ownersOwnerIsClusterGitTrack(e.Meta.GetOwnerReferences())
+}
+
+// Update returns true if the event object owner's owner is a ClusterGitTrack
+func (p OwnersOwnerIsClusterGitTrackPredicate) Update(e event.UpdateEvent) bool {
+	return p.ownersOwnerIsClusterGitTrack(e.MetaNew.GetOwnerReferences())
+}
+
+// Delete returns true if the event object owner's owner is a ClusterGitTrack
+func (p OwnersOwnerIsClusterGitTrackPredicate) Delete(e event.DeleteEvent) bool {
+	return p.ownersOwnerIsClusterGitTrack(e.Meta.GetOwnerReferences())
+}
+
+// Generic returns true if the event object owner's owner is a ClusterGitTrack
+func (p OwnersOwnerIsClusterGitTrackPredicate) Generic(e event.GenericEvent) bool {
+	return p.ownersOwnerIsClusterGitTrack(e.Meta.GetOwnerReferences())
+}
+
+func (p OwnersOwnerIsClusterGitTrackPredicate) ownersOwnerIsClusterGitTrack(ownerRefs []metav1.OwnerReference) bool {
+	gtoList := &farosv1alpha1.GitTrackObjectList{}
+	err := p.client.List(context.TODO(), gtoList)
+	if err != nil {
+		// We can't list GTOs so fail closed and ignore the requests
+		return false
 	}
+
+	// build a set of uids
+	gtoSet := make(map[types.UID]farosv1alpha1.GitTrackObject)
+	for _, item := range gtoList.Items {
+		gtoSet[item.UID] = item
+	}
+
+	for _, ref := range ownerRefs {
+		if ref.APIVersion != farosGroupVersion {
+			continue
+		}
+		// ClusterGitTrackObjects can only be owned by ClusterGitTracks, so we
```

Comment:
This sentence doesn't currently make sense, should work if you drop the we at the end of this line